nselib/libssh2-utility.lua

```diff
@@ -0,0 +1,172 @@
+---
+-- Utility functions for libssh2.
+--
+-- Provides helper class for the libssh2 binding that abstracts away 
+-- details of running remote commands.
+--
+-- @copyright Same as Nmap--See https://nmap.org/book/man-legal.html
+-- @class module
+-- @name libssh2-utility
+
+local stdnse = require "stdnse"
+
+local libssh2 = stdnse.silent_require "libssh2"
+
+SSHConnection = {}
+
+---
+-- Returns a new connection object.
+--
+-- @return A connection object.
+function SSHConnection:new()
+  local o = {}
+  setmetatable(o, self)
+  self.__index = self
+  return o
+end
+
+---
+-- Sets up a connection with a server.
+--
+-- @param host A host to connect to.
+-- @param port A port to connect to.
+-- @return true on success or nil on failure
+function SSHConnection:connect(host, port)
+  self.session = libssh2.session_open(host, port.number)
+  if self.session then
+    return true
+  end
+end
+
+
+---
+-- Sets up a connection with a server.
+--
+-- @param host A host to connect to.
+-- @param port A port to connect to.
+-- @return true on success or nil on failure
+function SSHConnection:connect_pcall(host, port)
+  local status, err 
+  status, self.session, err = pcall(libssh2.session_open, host, port.number)
+  return status, err
+end
+
+---
+-- Runs a shell command on the remote host.
+--
+-- @param cmd A command to run.
+-- @return The command output.
+function SSHConnection:run_remote(cmd)
+  if not (self.session and self.authenticated) then
+    return false
+  end  
+  local channel = libssh2.open_channel(self.session)
+  libssh2.channel_exec(self.session, channel, cmd) 
+  libssh2.channel_send_eof(self.session, channel)
+  local buff = ""
+  local data = ""
+  while not libssh2.channel_eof(channel) do
+    data = libssh2.channel_read(self.session, channel)
+    if data then
+      buff = buff .. data
+    end
+  end
+  return buff
+end
+
+---
+-- Attempts to authenticate using provided credentials.
+--
+-- @param username A username to authenticate as.
+-- @param password A password to authenticate as.
+-- @return true on success or false on failure.
+function SSHConnection:password_auth(username, password)
```

Comment:
Don't worry about it until this PR is merged first. But the goal would be for `ssh-brute` to use either one transparently, since many users don't/can't perceive a difference. Logic would look something like:

1. If password auth is supported, do that.
2. If not, but keyboard-interactive auth is supported, try to see if it's a password prompt.
3. If it's a password prompt, continue brute forcing over keyboard-interactive.

Keyboard-interactive can be other things, though, so maybe we'll have to add different script-args to enable other kinds of prompts, but the most common is a simple PAM password prompt. I'd guess (though haven't researched that part of the protocol) that the API would be something like handing the script a socket connected to the other end, and let the script do prompt detection (like telnet-brute does) and actual authentication.

Another fun thing might be a script to grab the "banner" of keyboard-interactive auth (OpenSSH's sshd_config uses the `Banner` directive for this).